mne/preprocessing/_csd.py

```diff
@@ -0,0 +1,197 @@
+#          Alex Rockhill <aprockhill206@gmail.com>
+#
+# License: BSD (3-clause)
+
+import numpy as np
+
+from scipy.linalg import inv
+
+from mne import pick_types, pick_info
+from mne.utils import logger
+from mne.io import BaseRaw
+from mne.epochs import BaseEpochs
+from mne.evoked import Evoked
+from mne.parallel import parallel_func
+from mne.channels.interpolation import _calc_g, _calc_h
+
+
+def _prepare_G(G, lambda2):
+    G.flat[::len(G) + 1] += lambda2
+    # compute the CSD
+    Gi = inv(G)
+
+    TC = Gi.sum(0)
+    sgi = np.sum(TC)  # compute sum total
+
+    return Gi, TC, sgi
+
+
+def _compute_csd(data, G_precomputed, H, head):
+    """compute the CSD"""
+    n_channels, n_times = data.shape
+    mu = data.mean(0)[None]
+    Z = data - mu
+    X = np.zeros_like(data)
+    head **= 2
+
+    Gi, TC, sgi = G_precomputed
+
+    Cp2 = np.dot(Gi, Z)
+    c02 = np.sum(Cp2, axis=0) / sgi
+    C2 = Cp2 - np.dot(TC[:, None], c02[None, :])
+    X = np.dot(C2.T, H).T / head
+    return X
+
+
+def compute_current_source_density(inst, picks=None, g_matrix=None,
+                                   h_matrix=None, lambda2=1e-5,
+                                   stiffness=4, n_legendre_terms=50,
+                                   head=1.0, n_jobs=1, copy=True):
+    """ Current Source Density (CSD) transformation
+
+    Transformation based on spherical spline surface Laplacian as suggested by
+    Perrin et al. (1989, 1990), published in appendix of Kayser J, Tenke CE,
+    Clin Neurophysiol 2006;117(2):348-368)
+
+    Implementation of algorithms described by Perrin, Pernier, Bertrand, and
+    Echallier in Electroenceph Clin Neurophysiol 1989;72(2):184-187, and
+    Corrigenda EEG 02274 in Electroenceph Clin Neurophysiol 1990;76:565.
+
+    Note: This function applies an average reference to the data.
+          Do not transform CSD data to source space.
+
+    Parameters
+    ----------
+    inst : instance of Raw, Epochs or Evoked
+        The data to be transformed.
+    picks : np.ndarray | list, shape (n_channels,) | None
+        The picks to be used. Defaults to None.
+    lambda2 : float
+        Regularization parameter, produces smoothnes. Defaults to 1e-5.
+    stiffnes : float
+        stiffnes of the spline. Also referred to as `m`
+        (if g_matrix or h_matrix is provided this will be ignored)
+    n_legendre_terms : int
+        number of Legendre terms to evaluate (if g_matrix or h_matrix
+        is provided this will be ignored)
+    head : float
+        The head radius (unit sphere). Defaults to 1.
+    g_matrix : ndarray, shape (n_channels, n_channels) | None
+        The matrix oncluding the channel pairwise g function. If None,
+        the g_function will be computed from the data (default).
+    h_matrix : ndarray, shape (n_channels, n_channels) | None
+        The matrix oncluding the channel pairwise g function. If None,
+        the h_function will be computed from the data (default).
```

Comment:
Let's remove these two pass-in-the-matrix options. We can always add them later.